src/main.cpp

```diff
@@ -5808,13 +5786,40 @@ bool SendMessages(CNode* pto)
             }
             pto->vInventoryBlockToSend.clear();
 
-            // Determine transactions to relay
-            bool fSendTrickle = pto->fWhitelisted;
-            if (pto->nNextInvSend < nNow) {
-                fSendTrickle = true;
-                // Use half the delay for outbound peers, as there is less privacy concern for them.
-                pto->nNextInvSend = PoissonNextSend(nNow, INVENTORY_BROADCAST_INTERVAL >> !pto->fInbound);
+            // Respond to BIP35 mempool requests
+            if (fSendTrickle && pto->fSendMempool) {
+                std::vector<uint256> vtxid;
+                mempool.queryHashes(vtxid);
+                pto->fSendMempool = false;
+
+                LOCK(pto->cs_filter);
+
+                BOOST_FOREACH(const uint256& hash, vtxid) {
+                    CInv inv(MSG_TX, hash);
+                    if (pto->pfilter) {
+                        CTransaction tx;
+                        bool fInMemPool = mempool.lookup(hash, tx);
+                        if (!fInMemPool) continue; // another thread removed since queryHashes, maybe...
+                        if (!pto->pfilter->IsRelevantAndUpdate(tx)) continue;
+                    }
+                    if (pto->minFeeFilter) {
+                        CFeeRate feeRate;
+                        mempool.lookupFeeRate(hash, feeRate);
+                        LOCK(pto->cs_feeFilter);
+                        if (feeRate.GetFeePerK() < pto->minFeeFilter)
+                            continue;
+                    }
+                    if (pto->vInventoryTxToSend.count(hash)) continue;
+                    pto->filterInventoryKnown.insert(hash);
+                    vInv.push_back(inv);
+                    if (vInv.size() == MAX_INV_SZ) {
+                        pto->PushMessage(NetMsgType::INV, vInv);
+                        vInv.clear();
+                    }
+                }
             }
+
+            // Determine transactions to relay
             if (fSendTrickle) {
```

Comment:
"&& vInv.size() < INVENTORY_BROADCAST_MAX" perhaps? otherwise when a mempool request has prefilled the INV it will waste time heapifying for no reason.